first pass at procedures (#105)
kwilteam/docs: docs/kuneiform/foreign_procedures.mdx

---
sidebar_position: 4
sidebar_label: "Foreign Procedures"
id: foreign-procedures
title: Foreign Procedures in Kuneiform
description: Using Foreign Procedures in Kuneiform
slug: /kuneiform/foreign-procedures
---

Foreign procedures are a way to define an interface for interacting with procedures in another database.
Foreign procedures can define an expected input and output, and specify the database and procedure that matches the interface.
The target database can be set dynamically during runtime, allowing for dynamic procedure calls across databases.

```typescript
database social_network;

// get_user is capable of calling a procedure in any other database on the network
// that takes a uuid and returns a text and an int
foreign procedure get_user($id uuid) returns (text, int)

// get_users is capable of calling a procedure in any other database on the network
// that takes no arguments and returns a table with columns username and age
foreign procedure get_users() returns table(username text, age int)
```

## Calling Foreign Procedures

> Foreign procedures can be called like any other function/procedure. The only difference is that the target

Reviewer: ```suggestion
Foreign procedures are called like any other function/procedure. The only difference is that the target
```

database id (dbid) and procedure name must be passed using square brackets:

```typescript
foreign procedure get_user($id uuid) returns (text, int)

procedure call_get_user($id uuid) public  {
    $username, $age = get_user['x_target_dbid', 'target_procedure']($id);
    // $username is a text and $age is an int
}
```

> Since the target procedure name is passed as a string, it is only necessary that the arguments and returns

Reviewer: wdym by "only"?

```suggestion
Since the target procedure name is passed as a string, it is necessary that the arguments and returns
```
